metrics/sinks/influxdb/influxdb_historical.go

```diff
@@ -0,0 +1,550 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package influxdb
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/heapster/metrics/core"
+
+	"github.com/golang/glog"
+	influxdb "github.com/influxdata/influxdb/client"
+	influx_models "github.com/influxdata/influxdb/models"
+)
+
+// Historical indicates that this sink supports being used as a HistoricalSource
+func (sink *influxdbSink) Historical() core.HistoricalSource {
+	return sink
+}
+
+// implementation of HistoricalSource for influxdbSink
+
+// Kube pod and namespace names are limitted to [a-zA-Z0-9-.], while docker also allows
+// underscores, so only allow these those characters.  When Influx actually supports bound
+// paramaters, this will be less necessary.
+var nameAllowedChars = regexp.MustCompile("^[a-zA-Z0-9_.-]+$")
+
+// sanitizeKey errors out if invalid characters are found in the key, since InfluxDB does not widely
+// support bound parameters yet (see https://github.com/influxdata/influxdb/pull/6634) and we need to
+// sanitize our inputs.
+func (sink *influxdbSink) checkSanitizedKey(key *core.HistoricalKey) error {
+	if key.NodeName != "" && !nameAllowedChars.MatchString(key.NodeName) {
+		return fmt.Errorf("Invalid node name %q", key.NodeName)
+	}
+
+	if key.NamespaceName != "" && !nameAllowedChars.MatchString(key.NamespaceName) {
+		return fmt.Errorf("Invalid namespace name %q", key.NamespaceName)
+	}
+
+	if key.PodName != "" && !nameAllowedChars.MatchString(key.PodName) {
+		return fmt.Errorf("Invalid pod name %q", key.PodName)
+	}
+
+	// NB: this prevents access to some of the free containers with slashes in their name
+	// (e.g. system.slice/foo.bar), but the Heapster API seems to choke on the slashes anyway
+	if key.ContainerName != "" && !nameAllowedChars.MatchString(key.ContainerName) {
+		return fmt.Errorf("Invalid container name %q", key.ContainerName)
+	}
+
+	if key.PodId != "" && !nameAllowedChars.MatchString(key.PodId) {
+		return fmt.Errorf("Invalid pod id %q", key.PodId)
+	}
+
+	return nil
+}
+
+// aggregationFunc converts an aggregation name into the equivalent call to an InfluxQL
+// aggregation function
+func (sink *influxdbSink) aggregationFunc(aggregationName core.AggregationType, fieldName string) string {
+	switch aggregationName {
```

Comment:
`aggregationName` values are checked against a fixed list at the API level (see `getAggregations` in `metrics/api/v1/historical_handlers.go:694`)